Add `program.json` + Support for Agent-Friendly CLI Input via stdio @@W-18855311@@

bendvc opened the pull request:
### Description

This PR introduces several improvements to the `create-mobify-app` script that make it easier to automate and integrate with AI agents or other tooling. Key updates include:

* **Introduced `program.json`**: A new JSON-based manifest that defines the behavior and structure of the CLI tool. This serves as a machine-readable contract for tools like the MCP server.
* **Added support for `stdio` input**: Enables passing answers via stdin, allowing agents to gather and pipe input directly into the CLI—great for non-interactive environments and automation.
* **Refactored internal CLI logic** to read from `program.json`, making the tool’s structure more declarative and extensible.
* **Updated data structures and templates** to align with this new configuration-driven model.
* **Added --displayProgram argument**: A new script action to display the schemas and data that this script uses, we ingest this from out MCP server.

These changes pave the way for more flexible, agent-driven CLI workflows while keeping the experience familiar for human users.

---

### Types of Changes

* [x] **New feature** (non-breaking change that adds functionality)
* [ ] Bug fix
* [ ] Documentation update
* [ ] Breaking change
* [ ] Other changes

---

### How to Test-Drive This PR

1. Pull the branch and run the CLI tool manually to ensure the prompts still behave as expected.
2. Pipe a JSON payload of answers into the CLI via stdin to verify agent-style automation works:

   ```bash
   echo '{"project.extend":true,"project.name":"demo-storefront","project.commerce.instanceUrl":"https://zzte-053.dx.commercecloud.salesforce.com","project.commerce.clientId":"1d763261-6522-4913-9d52-5d947d3b94c4","project.commerce.isSlasPrivate":false,"project.commerce.siteId":"RefArch","project.commerce.organizationId":"f_ecom_zzte_053","project.commerce.shortCode":"kv7kzm78","general.presetOrTemplateId":"retail-react-app"}' | node ./scripts/create-mobify-app --stdio
   ```
3. Check that updates to the generated app files (e.g., templates) reflect the expected changes.
4. Review `program.json` and confirm it correctly describes available questions and options.

---

### Checklists

#### General

* [ ] Changes are covered by test cases
* [ ] `CHANGELOG.md` updated with a short description of changes

#### Accessibility Compliance

* [x] There are no changes to UI

bendvc committed:
Initial Commit

- Create `program.json` to define project operation (this is consumed by the MCP server)
- Allow stdio input for answers. This allows integration with agents where they can gather answers and pipe them to the cli for project generation.
- Updates to data structure and asset templates.
- Use program.json in the script to configure the internal commander app

cc-prodsec commented:
![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/pr-banner-default.svg)

### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/a23ee650-6d82-44ee-a0f0-72c4f49d6bff?refs=b1c0f39239e5b85dcb6b1ff0292dbac34a1efa72&source=prChecksComment)

:white_check_mark: **license/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/a23ee650-6d82-44ee-a0f0-72c4f49d6bff/license?refs=b1c0f39239e5b85dcb6b1ff0292dbac34a1efa72&source=prChecksComment)

bendvc committed:
Update CHANGELOG.md

bendvc committed:
Ensure we include program.json

bendvc committed:
Update validation for url

bendvc committed:
Fix regression

Tried to use the package name as the template id, but this doesn't work with a lot of our current logic. So I reverted it.

vmarta commented on packages/pwa-kit-create-app/scripts/create-mobify-app.js:
```diff
@@ -851,71 +449,109 @@ const main = async (opts) => {
 
-    // Add the preset to the context.
-    context.preset = selectedPreset
+        // NOTE: This is a little weird, but we'll set this value to the template id and treat is as such from this point forward..
+        answers.general.presetOrTemplateId = selectedPreset.templateId
```
There are a few comments in this file where you said this may be weird.. how about merging both presets and templates into a single list?

AI shows what that could look like:

```js
const ALL_TEMPLATES = [
    // Convert presets to templates with default answers
    ...RAW_PRESETS.map((preset) => {
        const baseTemplate = RAW_TEMPLATES.find((t) => t.id === preset.templateId)
        if (!baseTemplate) {
            throw new Error(
                `Preset "${preset.id}" references unknown template "${preset.templateId}"`
            )
        }

        return {
            ...baseTemplate, // Template structure (questions, assets, etc.)
            id: preset.id, // Use preset ID as the unified ID
            name: preset.name, // Use preset name
            description: preset.description,
            shortDescription: preset.shortDescription,
            isPreset: true, // Flag to identify presets
            templateId: preset.templateId, // Keep reference to base template
            defaultAnswers: preset.answers, // Pre-filled answers
            private: preset.private
        }
    }),
    // Regular templates (no default answers)
    ...RAW_TEMPLATES.map((template) => ({
        ...template,
        isPreset: false,
        templateId: template.id, // Template references itself
        defaultAnswers: {} // No pre-filled answers
    }))
] 
```

Here's the full file: https://gist.github.com/vmarta/2273ea2dd094f1fab1df4f6eafe9ac82

bendvc replied:
I see there they are going with this, but I'm not sure it is in fact any less weird. I say that because this notion of "defaultAnswers". Initially if you had no idea about presets you would say that `defaultAnswers` would be used to show the user what the answer would be if they skipped the input or maybe use it as an example, but it would still ask you the question. 

But I'm assuming what we would have to do here is say, if the template is actually a preset, then when we ask the questions, use the defaultAnswers as provided answers so they won't get asked in the first place. 

And at first glance having a prop on a template called `isPreset` is still kinda weird, that data structure probably isn't a template but we use a variable called `all_templates` which serves to make us believe we are working with all templates.

How strongly do you feel about making this logic perfect before merging?

bendvc committed:
Fix npm view command

bendvc committed:
More small fixes

Related to change in the schema for the template

bendvc committed:
Fix more template references

adamraya commented on packages/pwa-kit-create-app/scripts/create-mobify-app.js:
```diff
@@ -851,71 +449,108 @@ const main = async (opts) => {
-        context.answers = await prompt(PRESET_QUESTIONS)
+    // If there is no preset provided via the CLI, check for stdio input or prompt the user
+    if (stdio) {
+        try {
+            const input = await readStdin()
+            answers = merge(answers, expandObject(JSON.parse(input)))
+        } catch (err) {
+            console.error('Failed to read from stdin:', err.message)
+            process.exit(1)
+        }
```
Should we split the code into separate steps to let users know which step fails by providing more specific error messages?

Also, should we check if the input includes all required fields?

Something like:
```
try {
        const input = await readStdin()
        if (!input.trim()) {
            throw new Error('No input received. Please pipe valid JSON to stdin.')
        }
        
        const parsedInput = JSON.parse(input)
        
        if (!parsedInput['general.presetOrTemplateId']) {
            throw new Error('Missing required field: "general.presetOrTemplateId"')
        }
        answers = merge(answers, expandObject(parsedInput))
    } catch (err) {
        if (err instanceof SyntaxError) {
            console.error('Invalid JSON format in stdin input')
        } else {
            console.error('Failed to process stdin input:', err.message)
        }
        process.exit(1)
}
```

bendvc replied:
Thats a good called out. I can organize the code a little bit more to better show errors. But I'm not too sure about the answer validation in it's entirety... Maybe we can go with what you have here and validate that there is a presetORTemplateId.. the other answers are template specific and maybe we can create a follow up ticket to implement the stdio input better.

bendvc committed:
PR feedback

adamraya commented on packages/pwa-kit-create-app/scripts/create-mobify-app.js:
```diff
@@ -851,71 +449,108 @@ const main = async (opts) => {
+        return {
+            ...question,
+            validate: validator?.regex
+                ? (input) => new RegExp(validator.regex, 'i').test(input) || validator.message
+                : undefined
```
Should we catch errors in case the `validator.regex` is invalid?

bendvc replied:
In the future I think there will be a schema validation step that could be used to do something like this. Right now this is all a design time and if there is an error it's because the pwa-kit developer added something wrong to the program json and it's not something that would effect external developers that are using the script.

bendvc committed:
Print program data to be consumed by the mcp server or other ai

adamraya commented on packages/pwa-kit-create-app/scripts/create-mobify-app.js:
```diff
@@ -739,30 +308,30 @@ const runGenerator = (context, {outputDir, templateVersion, verbose}) => {
     }
 
     // Extract the source
+    console.log('Extracting base template from package or npm: ', tarPath)
```
Are we intentionally leaving this console.log? Should we print this behind a verbose or debug flag?

bendvc replied:
No. Thanks for spotting that. I have removed it.

bendvc committed:
Remove debug log

vmarta commented on packages/pwa-kit-create-app/scripts/create-mobify-app.js:
```diff
@@ -414,6 +413,66 @@ const readStdin = async () => {
+ * @param {Object} answers - The current answers object to merge into.
+ * @returns {Promise<Object>} - The merged answers object.
+ */
+const getAnswersFromStdin = async (answers) => {
```
Minor: I feel the function name doesn't imply that there's going to be a merging of the initial answers and the ones from stdin. So when I see line 518 `answers = await getAnswersFromStdin(answers)`, it made me think why we need to pass in `answers`.

How about doing the answers merging outside of this function? The error handling is focusing on json and stdin anyways, so it looks to me like the merging can be done outside.

vmarta reviewed (dismissed):
A minor comment.. other than that, PR looks good to me 👍

adamraya commented on packages/pwa-kit-create-app/scripts/create-mobify-app.js:
```diff
@@ -819,6 +388,91 @@ const foundNode = process.versions.node
+    } catch (err) {
+        console.error('Failed to import program.json:', err.message)
+    }
+    process.exit(1)
```
Why are we using `process.exit(1)` for a successful operation? Should we use  `process.exit(0)` instead? Or move it inside the catch block?

bendvc replied:
0 is success from what the ide is telling me.

adamraya commented on packages/pwa-kit-create-app/scripts/create-mobify-app.js:
```diff
@@ -834,11 +488,19 @@ const main = async (opts) => {
     const OUTPUT_DIR_FLAG_ACTIVE = !!outputDir
     const presetId = preset || process.env.GENERATOR_PRESET
 
+    // Exit if the preset provided is not valid.
```
nit: Is this comment copied and pasted? Should it instead say something like: "Display program schema and exit if requested"?

bendvc committed:
PR feedback

bendvc committed:
Fix AI code
